[FSDP] AutoWrap Main API

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68776
* __->__ #68155

Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma committed:
[FSDP] AutoWrap Main API

Per title

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements 
Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394


Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.


Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -45,86 +235,133 @@ def test_wrap_override_defaults(self):
 
     @unittest.skipIf(not torch.cuda.is_available(), "Test Requires CUDA")
-    def test_auto_wrap_smoke_test(self):
+    @parametrize("wrap_method", [WrapMethod.FSDP_CTOR])
```
will add the other one too

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -133,14 +370,20 @@ def test_auto_wrap_smoke_test(self):
-                my_auto_wrap_policy = functools.partial(default_auto_wrap_policy, min_num_params=40)
-                model = auto_wrap(sequential, auto_wrap_policy=my_auto_wrap_policy)
-            model.to(device)
+            sequential = TestFSDPWrap.NestedSequentialModel.get_model(cuda=False)
```
probably useful to test with CUDA before and CUDA after as well, though afaik CUDA after is a bit more important, as the intial model might fit on CPU but not GPU.

zhaojuanmao replied:
sounds good

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -45,86 +235,133 @@ def test_wrap_override_defaults(self):
     @unittest.skipIf(not torch.cuda.is_available(), "Test Requires CUDA")
-    def test_auto_wrap_smoke_test(self):
+    @parametrize("wrap_method", [WrapMethod.FSDP_CTOR])
+    def test_auto_wrap_smoke_test(self, wrap_method):
```
also need to test it with cpu offload

rohan-varma commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -293,7 +303,7 @@ def recursive_wrap(module: nn.Module, auto_wrap_policy: Optional[Callable], **kw
             # since the left over parameters exceed the number of params to wrap
             remainder = num_params - total_wrapped_params
-            if auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder):
+            if not only_wrap_children and auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder):
```
recursive_wrap by default would wrap the passed in module as well, but we already called FSDP(model, policy) on it, it will cause issues if we do not skip it like this

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping.
- Figure out what to do if we call `FSDP(model, policy=...)` and the outer-most model should not be wrapped according to `policy`. Currently it is still FSDP because we call the constructor. We can either warn or throw here if the outermost model does not satisfy `policy`. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
             will be no offloading.
+        auto_wrap_policy: (Optional [callable]):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+            Note that this policy currently will only apply to child modules of
```
> If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

Could you please elaborate on the above statement? What's missing for the outer module? Do you mean the parameters directly created from the outer module?

mrshenli replied:
Does it mean we allow users to pass an FSDP module into FSDP ctor?

rohan-varma replied:
I meant that current `wrap` API supports something like:
```
with enable_wrap(policy):
    wrap(mod) # assume mod, outer-most module does not satisfy policy
```

then `mod` will stay as an nn.Module, but with this new API, we have like - 
```
mod = FSDP(mod, policy)
```
and `policy` is currently only applied to children. It is tricky to apply `policy` in this way to `mod` itself because it will still be FSDP, since we call the ctor. 

> Does it mean we allow users to pass an FSDP module into FSDP ctor?

We disallow this, there is the function `_check_wrapped` to guard against it

zhaojuanmao replied:
in real use cases, users always need to do FSDP(auto_wrap(module)), that means we always need an outermost/root FSDP instance for a module. so it is good to always return a root FSDP instance in the new main API FSDP(module, auto_wrap_policy)

rohan-varma replied:
@mrshenli Does this explanation sound good to you?

mrshenli replied:
I see, this makes sense to me. Thanks!

mrshenli commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
         process_group: Optional[ProcessGroup] = None,
-        cpu_offload: Optional[CPUOffload] = None
+        cpu_offload: Optional[CPUOffload] = None,
+        fsdp_auto_wrap_policy: Optional[callable] = None,
```
this disagrees with the doc `auto_wrap_policy`

mrshenli commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -254,7 +254,12 @@ def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        auto_wrap_policy: Optional[Callable],
+        only_wrap_children: bool = False,
+        **kwargs: Any
+        ) -> Tuple[nn.Module, int]:
```
lint error is real

mrshenli commented on torch/distributed/_fsdp/wrap.py:
```diff
@@ -264,6 +269,11 @@ def recursive_wrap(module: nn.Module, auto_wrap_policy: Optional[Callable], **kw
                 module to recursively wrap
             auto_wrap_policy (Callable, Optional):
                 optionally, override the :func:`auto_wrap_policy` from the context.
+            only_wrap_children (bool, Optional):
```
nit: curious, under the current plan, what belongs to `auto_wrap_policy`, and what should be separate args like `only_wrap_children`?

rohan-varma replied:
this arg is for private `ConfigAutoWrap._recursive_wrap` API which should not be used directly by users. So under the current plan, users only have to pass in the policy, something like

```
module = FSDP(module, auto_wrap_policy=partial(default_policy, min_num_params=1e7))
```

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -23,6 +23,8 @@
 from torch.nn.parameter import Parameter
 
 from .flatten_params_wrapper import FlattenParamsWrapper
+from .wrap import ConfigAutoWrap, auto_wrap
```
we do not need auto_wrap?

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
+        auto_wrap_policy: (Optional [callable]):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+            Note that this policy currently will only apply to child modules of
+            the passed in module. The overall module is always FSDP because we
```
nit:  the passed in module. The reminder modules are always wrapped in the returned root FSDP instance

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
+                wrapped_module, total_wrapped_params = ConfigAutoWrap.recursive_wrap(
+                    module,
+                    auto_wrap_policy=fsdp_auto_wrap_policy,
+                    # Note that we have the recursive_wrap skip wrapping for
+                    # the outermost (this) module otherwise it will result in a
+                    # double-wrap causing issues.
+                    only_wrap_children=True,
```
it is OK to wrap outermost module here?  we can check after recursive_wrap(): if total_wrapped_params== num_params, it can early return in line 148?

rohan-varma replied:
Discussed offline with @zhaojuanmao, agreed that we will go with the current approach as returning early will still result in a double wrap, will `self.module` will become the true FSDP instance, but the user gets back the `self` instance. Basically we need to ensure that the remainder of the ctor logic runs for the root module.

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
+                check_fn=lambda mod: not isinstance(mod, FullyShardedDataParallel),
+                err_fn=lambda mod: f"Expected {mod} to NOT be FullyShardedDataParallel if auto_wrap is enabled.",
+            )
+            # TODO: refactor recursive_wrap so that it is not dependent on
+            # ConfigAutoWrap.
```
agree with this, we need to refactor this in the follow up PRs

rohan-varma replied:
Yep, will remove it in follow up PRs.

zhaojuanmao commented:
Thanks for the work!  

In terms of follow up items, I'm thinking as follows: 

1.  continue to support "enable_wrap() and wrap()" API, this provides a convenient way to pass same configs for all nested manual wrapping, but we need to refactor and simply the APIs, just give a context and pass same configs for all wrap() functions, not need to pass auto_wrap_policy and etc.  

we still have to support manual wrapping for use cases until auto_wrap_policy (through ) is easy/efficient to use and can cover most of use cases.  

mostly at the beginning users do not know the optimized wrapping strategy until they manually wrapped and profiled it. we can provide a list of best-practice wrapping policies, or profiling tools or auto tuning support in the future to help users. But before, I think we need to provide this convenient APIs for users for manual wrapping.  I will put 'best-practice wrapping policies/profiling tools to find best sharding strategies/auto tuning' into our roadmap. 

2.  we can kill auto_wrap() API, instead encourage users to use this new main API with auto_wrapping functionality. As it will be easier for meta device support in the future. also 'FSDP(module, auto_wrap_policy)' is a natural replacement of current usage 'FSDP(auto_wrap(module))'. 

3. agree that we should not let _recursive_wrap depend on ConfigAutoWrap context manager, as there is only one main FSDP API, no need context manger here probably

4.  agree that we need to update callsite of FSDP API to use auto_wrapping as much as possible, if the auto_wrap_policy can not satisfy the use cases, use the refactored manual wrap API

5. agree that we need to make it easier to use policies for wrapping etc. i.e. provide a wrap_policies class with functions for common policies (such as # of parameters). like mentioned 'best-practice wrapping policies/profiling tools to find best sharding strategies/auto tuning' into our roadmap

zhaojuanmao commented:
regarding best practice auto wrapping policies, e..g, we can have a way to make sure outer most reminder modules to be None or last layers, otherwise it will consume more memories. We can add more once we study more use cases in the future.

rohan-varma commented:
After discussion with @zhaojuanmao we decided to keep `only_wrap_children` in private `_recursive_wrap` API for now as the # of parameter checking approach would only work if we weren't already in FSDP constructor. Since we are wrapping outermost module in FSDP constructor, it will be the one that is responsible for wrapping outermost module. 

Follow up changes will - 

1. Make recursive_wrap not depend on ConfigAutoWrap
2. Kill auto_wrap, it is not used and causes extra user confusion. Manual wrapping APIs we will support are enable_wrap and wrap which allows user to manually wrap FSDP modules.

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -1,34 +1,224 @@
+        model = MyModel(nested=nested)
+        return model
+
+    # TODO: this generates 8 tests, is it worth testing all these config?
```
sounds good, only keeping nested parameter for this test for now then.

zhaojuanmao replied:
nit: remove this?

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping.
- Figure out what to do if we call `FSDP(model, policy=...)` and the outer-most model should not be wrapped according to `policy`. Currently it is still FSDP because we call the constructor. We can either warn or throw here if the outermost model does not satisfy `policy`. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping.
- Figure out what to do if we call `FSDP(model, policy=...)` and the outer-most model should not be wrapped according to `policy`. Currently it is still FSDP because we call the constructor. We can either warn or throw here if the outermost model does not satisfy `policy`. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping.
- Figure out what to do if we call `FSDP(model, policy=...)` and the outer-most model should not be wrapped according to `policy`. Currently it is still FSDP because we call the constructor. We can either warn or throw here if the outermost model does not satisfy `policy`. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping.
- Figure out what to do if we call `FSDP(model, policy=...)` and the outer-most model should not be wrapped according to `policy`. Currently it is still FSDP because we call the constructor. We can either warn or throw here if the outermost model does not satisfy `policy`. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -95,16 +97,60 @@ class FullyShardedDataParallel(nn.Module):
+                )
+            assert not ConfigAutoWrap.in_autowrap_context
+
+        # TODO: if we want to apply the policy to the outermost module, it
+        # should be applied here and we can have a flag to control whether this
+        # module behaves like an nn.Module or FSDP.
```
nit: remove?

zhaojuanmao reviewed (approved):
awesome work! just two minor nits

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] AutoWrap Main API"


Implements auto wrapping API as part of FSDP constructor, to make it easier for users to recursively wrap FSDP module. 
API proposal: https://github.com/pytorch/pytorch/issues/64394. 

Now, FSDP can be used as follows, with user providing a policy or using `default_auto_wrap_policy`: 

```
class model:
    def _init(self):
        self.layer1 = nn.linear()
        self.layer2 = nn.linear()
        self.layer3 = nn.linear()
fsdp_model = FSDP(model(), 
                  auto_wrap_policy=functools.partial(default_auto_wrap_policy, min_num_params=1e7))
```

note that `auto_wrap_policy` currently only applies to child modules, and the outer-most (i.e. module passed into FSDP constructor will always be wrapped with FSDP, and be an instance of FSDP since we call FSDP constructor on it). 

If we need more flexibility and run into situations where the outer module should also not be wrapped, we can have a workaround where we run the outer module like a normal nn.Module by providing a special mode for this.

## Follow up items (will file issues)
- Need to remove the dependency on ConfigAutoWrap and make _recursive_wrap work in a standalone, functional fashion, so that eventually we can remove most `wrap.py` code and consolidate one way to do wrapping. 
- Update other callsites in test code to use this wrap API
- Update lightning/other callsites --> this will help us uncover any shortcomings in the API.
- Make it easier to use policies for wrapping etc. i.e. provide a `wrap_policies` class with functions for common policies (such as # of parameters).

Differential Revision: [D32327954](https://our.internmc.facebook.com/intern/diff/D32327954/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
failure is unrelated - Cannot find source file:

    /var/lib/jenkins/workspace/aten/src/THC/THCStorage.cpp
